File: Windows/Power of God Lib/Utilities/Network.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Windows.Forms;
using Power_of_God_Lib.pSystem;

namespace Power_of_God_Lib.Utilities
{
    public class Network
    {
        
        private const string VersionPrefix = "Pre-Release";
        private const string CurrentPreRelease = "b1a003";

        

        private static readonly string[] CurrentVersion = {
            "1", "0", null, null
        };

        public static string LatestStable()
        {
            if (CurrentPrefixInt() != 5000)
            {
                var returnValue = VersionPrefix + " " + CurrentVersion[0] + "." +
                 CurrentVersion[1];
                if (CurrentVersion[2] == null) return returnValue;
                returnValue += "." + CurrentVersion[2];
                if (CurrentVersion[3] != null)
                {
                    returnValue += "." + CurrentVersion[3];
                }
                return returnValue;
            }
            else
            {
                return "Pre-Release " + CurrentPreRelease;
            }
            
        }

        private static int CurrentPrefixInt()
        {
            switch (VersionPrefix)
            {
                case "Alpha":
                    return 0;
                case "Beta":
                    return 1;
                case "Pre-Release":
                    return 5000;
                default:
                    return 2;
            }
        }

        public static string ServerResponse = "Nothing Yet";

        public static string UpdateNotice()
        {
            switch (UpdateWord())
            {
                case "Updated":
                    return "You are currently updated to the most recent version.";
                case "Outdated":
                    return "You do not have the most recent version. Consider updating to " +
                           LatestOnline();
                default:
                    if (CurrentPrefixInt() == 5000)
                    {
                        return
                            "You are using a Pre-Release. Please note that not all features will be functional.\n\n" +
                            "(" + LatestStable() + ")";
                    }
                    else
                    {
                        return "You are using an unsupported version of Power of God. \nPlease consider " +
                            "using the current version, " + LatestOnline() +
                            ", \nbecause your version could possibly be unstable.";
                    }
            }
        }

        public static string UpdateWord()
        {
            try

            {

                int onlinePrefix;
                switch (GetUrlSource(Url).ElementAt(0).Substring(1))
                {
                    case "Alpha":
                        onlinePrefix = 0;
                        break;
                    case "Beta":
                        onlinePrefix = 1;
                        break;
                    default:
                        onlinePrefix = 2;
                        break;
                }
                GetUrlSource(Url);
                var itemcontext = GetUrlSource(Url).ElementAt(1);
                var item1 = int.Parse(itemcontext);
                var item2 = int.Parse(GetUrlSource(Url).ElementAt(2));
                int item3; // If these values stay as -1 then the app will know to not read them
                var item4 = -1;
                try
                {
                    item3 = int.Parse(GetUrlSource(Url).ElementAt(3));
                    try
                    {
                        item4 = int.Parse(GetUrlSource(Url).ElementAt(4));
                    }
                    catch (Exception)
                    {
                        item4 = -1;
                    }
                }
                catch (Exception)
                {
                    item3 = -1;
                }
                if (onlinePrefix > CurrentPrefixInt()) return "Outdated";
                if (onlinePrefix < CurrentPrefixInt()) return "Unsupported";
                if (item1 > CurrentVersionInt().ElementAt(0)) return "Outdated";
                if (item1 < CurrentVersionInt().ElementAt(0)) return "Unsupported";
                if (item2 > CurrentVersionInt().ElementAt(1)) return "Outdated";
                if (item2 < CurrentVersionInt().ElementAt(1)) return "Unsupported";
                if (item3 > -1)
                {
                    var local3 = CurrentVersionInt().ElementAt(2);
                    if (item3 > local3) return "Outdated";
                    if (item3 < local3) return "Unsupported";
                    if (item4 > -1)
                    {
                        var local4 = CurrentVersionInt().ElementAt(3);
                        if (item4 > local4) return "Outdated";
                        if (item4 < local4) return "Unsupported";
                    }
                    else
                    {
                        var localHere = CurrentVersion[3];
                        if (localHere != null) return "Unsupported";
                    }
                }
                else
                {
                    var localHere = CurrentVersion[2];
                    if (localHere != null) return "Unsupported";
                }
                return "Updated"; // If it gets to this point - the user has passed all checks. They are updated!
            }
            catch (Exception ex)
            {
                ErrorLogging.Write(ex);
                return "Updated"; // Exception caught. Let the user continue.
            }
        }

        private static IEnumerable<int> CurrentVersionInt()
        {
            var xList = new List<int>();
            foreach (var x in CurrentVersion)
            {
                try
                {
                    xList.Add(int.Parse(x));
                }
                catch (Exception)
                {
                    xList.Add(-1);
                }
            }
            return xList;
        }

        private const string Url = "http://godispower.us/Application/Updates.txt";

        public static string LatestOnline()
        {
            var x = GetUrlSource(Url).ElementAt(0) + " ";
            var versionStoppedAt = 0;
            for (var x2 = 1; x2 <= 4; x2++)
            {
                if (x2 >= 5) continue;
                if (!GetUrlSource(Url).ElementAt(x2).Equals("NR"))
                {
                    if (x2 < 4) x += "" + GetUrlSource(Url).ElementAt(x2) + ".";
                    else if (x2 == 4) x += "" + GetUrlSource(Url).ElementAt(x2);
                }
                else
                {
                    versionStoppedAt = x2;
                }
            }
            if (versionStoppedAt == 3 || versionStoppedAt == 4)
            {
                return x.Substring(0, x.Length - 1);
            }
            return x;
        }

        public static List<string> GetUrlSource(string urlF)
        {
            if (urlF.Contains("0201"))
            {
                urlF = urlF.Replace("0201", "2016");
            }
            var temp = "power.of.god/check_file.txt";
            var c = File.CreateText(temp);

            c.Close();
            using (var client = new WebClient())
            {
                try
                {
                    client.DownloadFile(urlF, temp);
                }
                catch (Exception e)
                {
                    ErrorLogging.Write(e);
                }

            }
            return File.ReadAllLines(temp).ToList();
        }

        public static WebBrowser XBrowser = new WebBrowser();
        public static void UpdateWebContent(string html)
        {
            XBrowser.DocumentText = html;
        }

        public static void Navigate(string url)
        {
            XBrowser.Navigate(url);
        }

        public static string Mediafire(string download)
        {
            var req = (HttpWebRequest)WebRequest.Create(download);
            var res = (HttpWebResponse)req.GetResponse();
            // ReSharper disable once AssignNullToNotNullAttribute
            var str = new StreamReader(res.GetResponseStream()).ReadToEnd();
            var indexurl = str.IndexOf("http://download", StringComparison.Ordinal);
            var indexend = GetNextIndexOf('"', str, indexurl);
            var direct = str.Substring(indexurl, indexend - indexurl);
            return direct;
        }

        private static int GetNextIndexOf(char c, string source, int start)
        {
            if (start < 0 || start > source.Length - 1)
            {
                throw new ArgumentOutOfRangeException();
            }
            for (var i = start; i < source.Length; i++)
            {
                if (source[i] == c)
                {
                    return i;
                }
            }
            return -1;
        }

        public static void DownloadFile(string internetLocation, string localLocation)
        {
            using (var client = new WebClient())
            {
                client.DownloadFile(internetLocation, localLocation);
            }
        }
    }

    
}

File: Java/Power of God Lib/src/Utilities/Network.java
/*
 * The MIT License
 *
 * Copyright 2015 apotter96.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package Utilities;

import java.io.*;
import java.net.*;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.Scanner;
import java.util.zip.*;

/**
 *
 * @author apotter96
 */
public class Network {
    private static final String VersionPrefix = "Alpha";
    private static final String[] CurrentVersion = new String[]
    {
        "1", "5", null, null
    };
    private static String TogetherNumbers()
    {
        int VersionPrefixCount = VersionPrefix.length();
        return LatestStable().substring(VersionPrefixCount + 1);
    }
    public static String LatestStable()
    {
        String returnValue = VersionPrefix + " " + CurrentVersion[0] + "." + 
                CurrentVersion[1];
        if (CurrentVersion[2] != null)
        {
            returnValue += "." + CurrentVersion[2];
            if (CurrentVersion[3] != null)
            {
                returnValue += "." + CurrentVersion[3];
            }
        }
        return returnValue;
    }
    private static int CurrentPrefixInt()
    {
        switch (VersionPrefix)
        {
            case "Alpha":
                return 0;
            case "Beta":
                return 1;
            default:
                return 2;
        }
    }
    public static String ServerResponse = "Nothing Yet";
    //Checking version based on javascript execution is now legacy
    //reading based on txt file from web concept thanks to Horfius on #coders (irc.esper.net)
    public static String UpdateNotice() throws IOException
    {
        switch (UpdateWord())
        {
            case "Updated":
                return "You are currently updated to the most recent version";
            case "Outdated":
                return "You do not have the most recent version. \nYou should consider " + 
                        "updating to " + LatestOnline();
            default:
                return "You are using an unsupported version of Power of God. \nPlease consider " + 
                      "using the current version, " + LatestOnline() + 
                      ", \nbecause your version could possibly be unstable.";

        }
    }
    private static String UpdateWord() throws IOException
    {
        int OnlinePrefix;
        switch (getUrlSource(url).get(0).substring(1)) 
        {
            case "Alpha": OnlinePrefix = 0; break;
            case "Beta": OnlinePrefix = 1; break;
            default: OnlinePrefix = 2; break;
        }
        int Item1 = Integer.parseInt(getUrlSource(url).get(1));
        int Item2 = Integer.parseInt(getUrlSource(url).get(2));
        int Item3 = -1; // If these values stay as -1 then the app will know to not read them
        int Item4 = -1;
        try
        {        
            Item3 = Integer.parseInt(getUrlSource(url).get(3));
            try
            {
                Item4 = Integer.parseInt(getUrlSource(url).get(4));
            }
            catch (IOException | NumberFormatException ex)
            {
                Item4 = -1;
            }
        }
        catch (Exception ex)
        {
            Item3 = -1;
        }
        if (OnlinePrefix > CurrentPrefixInt()) return "Outdated";
        if (OnlinePrefix < CurrentPrefixInt()) return "Unsupported";
        if (Item1 > CurrentVersionInt().get(0)) return "Outdated";
        if (Item1 < CurrentVersionInt().get(0)) return "Unsupported";
        if (Item2 > CurrentVersionInt().get(1)) return "Outdated";
        if (Item2 < CurrentVersionInt().get(1)) return "Unsupported";
        if (Item3 > -1)
        {
            int local3 = CurrentVersionInt().get(2);
            if (Item3 > local3) return "Outdated";
            if (Item3 < local3) return "Unsupported";
            if (Item4 > -1)
            {
                int local4 = CurrentVersionInt().get(3);
                if (Item4 > local4) return "Outdated";
                if (Item4 < local4) return "Unsupported";
            }
            else
            {
                String localHere = CurrentVersion[3];
                if (localHere != null) return "Unsupported";
            }
        } 
        else
        {
            String localHere = CurrentVersion[2];
            if (localHere != null) return "Unsupported";
        }
        return "Updated"; // If it gets to this point - the user has passed all checks. They are updated!
    }
    private static ArrayList<Integer> CurrentVersionInt()
    {
        ArrayList<Integer> xList = new ArrayList<>();
        for (String x:CurrentVersion)
        {
            try
            {
                xList.add(Integer.parseInt(x));
            }
            catch (Exception ex)
            {
                xList.add(-1);
            }
        }
        return xList;
    }
    private static String url = "http://godispower.us/Application/Updates.txt";
    public static String LatestOnline() throws IOException
    {
        String x = getUrlSource(url).get(0) + " ";
        int VersionStoppedAt = 0;
        for (int x2 = 1; x2 <= 4; x2++)
        {
            if (x2 < 5)
            {
                if (!getUrlSource(url).get(x2).equals("NR"))
                {
                    if (x2 < 4) x += "" + getUrlSource(url).get(x2) + ".";
                    else if (x2 == 4) x += "" + getUrlSource(url).get(x2);
                }
                else
                {
                    VersionStoppedAt = x2;
                }
            }
        }
        if (VersionStoppedAt == 3 || VersionStoppedAt == 4)
        {
            return x.substring(0, x.length() - 1);
        }
        return x;
    }
    public static ArrayList<String> getUrlSource(String urlF) throws IOException {
        URL url = new URL(urlF);
        InputStream is = url.openStream();
        Scanner s = new Scanner(is, "UTF-8");
        ArrayList<String> fileLines = new ArrayList<>();
        while (s.hasNextLine())
        {
            fileLines.add(s.nextLine());
        }
        return fileLines;
    }
}
